src/Grpc.Net.Client/Balancer/Resolver.cs

```diff
@@ -0,0 +1,155 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Grpc.Net.Client.Configuration;
+
+namespace Grpc.Net.Client.Balancer
+{
+    /// <summary>
+    /// A configurable component that resolves a target <see cref="Uri"/> and returns them to the caller.
+    /// The targets URI's scheme is used to select the <see cref="Resolver"/> implementation, and uses the
+    /// URI parts after the scheme for actual resolution.
+    /// <para>
+    /// The addresses of a target may change over time, thus the caller registers a callback to receive
+    /// continuous updates as <see cref="ResolverResult"/>.
+    /// </para>
+    /// <para>
+    /// A <see cref="Resolver"/> doesn't need to automatically re-resolve on failure. Instead, the callback
+    /// is responsible for eventually invoking <see cref="RefreshAsync(CancellationToken)"/>.
+    /// </para>
+    /// </summary>
+#if HAVE_LOAD_BALANCING
+    public
+#else
+    internal
+#endif
+        abstract class Resolver : IDisposable
+    {
+        /// <summary>
+        /// Starts resolution.
+        /// </summary>
+        /// <param name="listener">The callback used to receive updates on the target.</param>
+        public abstract void Start(Action<ResolverResult> listener);
+
+        /// <summary>
+        /// Refresh resolution. Can only be called after <see cref="Start(Action{ResolverResult})"/>.
+        /// <para>
+        /// This is only a hint. Implementation takes it as a signal but may not start resolution.
+        /// </para>
+        /// </summary>
+        /// <param name="cancellationToken">A cancellation token.</param>
+        /// <returns>A task.</returns>
+        public abstract Task RefreshAsync(CancellationToken cancellationToken);
```

Comment:
When resolver refresh is requested we don't want to wait for the result from the calling code.

Load balancing is event driven (load balancer will be called once something has happened) rather than imperative (call a thing and wait for it to finish)